File: server/tools/buildTopology.ts
/**
 * Generate topology/*.json from the simulator's level layouts (settings/lvl*.json).
 *
 * Each entry/exit sensor is paired with its nearest gate (closest pairs first, one gate
 * per sensor, within --max-distance). Review the output - especially gate_distance -
 * and commit it; the app picks the matching file at startup.
 *
 *   npm run topology -- --levels-dir "C:/.../ParkingSimulator-win-x64/settings"
 */
import { mkdirSync, readdirSync, writeFileSync } from "node:fs";
import path from "node:path";
import { parseArgs } from "node:util";
import { loadDotEnv, loadSettings } from "../src/config";
import { deriveFromLevelFile } from "../src/topology";

loadDotEnv();
const cfg = loadSettings();
const { values } = parseArgs({
  options: {
    "levels-dir": { type: "string", default: cfg.simLevelsDir },
    out: { type: "string", default: cfg.topologyDir },
    "max-distance": { type: "string", default: String(cfg.topologyMaxGateDistance) },
  },
});
const levelsDir = values["levels-dir"];
if (!levelsDir) {
  console.error("--levels-dir is required (or set GPA_SIM_LEVELS_DIR)");
  process.exit(2);
}
const out = path.resolve(values.out!);
mkdirSync(out, { recursive: true });

for (const file of readdirSync(levelsDir).filter((f) => /^lvl.*\.json$/.test(f)).sort()) {
  const topo = deriveFromLevelFile(path.join(levelsDir, file), Number(values["max-distance"]));
  const { source, ...toWrite } = topo;
  const target = path.join(out, `${topo.name}.json`);
  writeFileSync(target, JSON.stringify(toWrite, null, 2) + "\n");
  const lanes = [...topo.entry_lanes, ...topo.exit_lanes];
  const unpaired = lanes.filter((l) => !l.gate).map((l) => l.spot);
  console.log(`${target}: ${topo.entry_lanes.length} entries, ${topo.exit_lanes.length} exits` +
    (unpaired.length ? `  WARNING no gate for ${unpaired.join(", ")}` : ""));
  const dist = topo.notes!.gate_distance as Record<string, number | null>;
  for (const l of lanes) console.log(`   ${l.spot.padEnd(10)} -> ${String(l.gate).padEnd(7)} zone=${(l.zone || "-").padEnd(6)} dist=${dist[l.spot]}`);
}

File: tools/build_topology.py
"""Generate topology/*.json from the simulator's level layouts (settings/lvl*.json).

Each entry/exit sensor is paired with its nearest gate (closest pairs first, one
gate per sensor, within --max-distance). Review the output - especially the
gate_distance notes - and commit it; the app picks the matching file at startup.

Run from the repo root:
  .venv\\Scripts\\python -m tools.build_topology --levels-dir "C:/.../ParkingSimulator-win-x64/settings"
"""
import argparse
import json
from pathlib import Path

from app.config import settings
from app.topology import derive_from_level_file


def main():
    ap = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("--levels-dir", type=Path, default=settings.sim_levels_dir, required=settings.sim_levels_dir is None,
                    help="simulator settings folder containing lvl*.json (default: GPA_SIM_LEVELS_DIR)")
    ap.add_argument("--out", type=Path, default=settings.topology_dir, help="output folder (default: GPA_TOPOLOGY_DIR)")
    ap.add_argument("--max-distance", type=float, default=settings.topology_max_gate_distance)
    args = ap.parse_args()

    args.out.mkdir(parents=True, exist_ok=True)
    for path in sorted(args.levels_dir.glob("lvl*.json")):
        topo = derive_from_level_file(path, args.max_distance)
        target = args.out / f"{topo.name}.json"
        target.write_text(json.dumps(topo.to_dict(), indent=2) + "\n", encoding="utf-8")
        unpaired = [l.spot for l in topo.entry_lanes + topo.exit_lanes if not l.gate]
        print(f"{target}: {len(topo.entry_lanes)} entries, {len(topo.exit_lanes)} exits"
              + (f"  WARNING no gate for {unpaired}" if unpaired else ""))
        for lane in topo.entry_lanes + topo.exit_lanes:
            print(f"   {lane.spot:10} -> {str(lane.gate):7} zone={lane.zone or '-':6} "
                  f"dist={topo.notes['gate_distance'][lane.spot]}")


if __name__ == "__main__":
    main()
